README.md
# convmodel

![](https://github.com/colorfulscoop/convmodel/workflows/unittest/badge.svg)

**convmodel** provides a conversation model based on [transformers](https://github.com/huggingface/transformers) GPT-2 model :wink:

:sparkles: Features :sparkles:

* Utilizes GPT2 model to generate response
* Handles multi-turn conversation
* Provides useuful interfaces to fine-tune model and generate a response from a given context

A simple example of fine-tune GPT-2 model and generate a response:

```py
from convmodel import ConversationModel
from convmodel import ConversationExample

# Load model on GPU
model = ConversationModel.from_pretrained("gpt2")

# Define training/validation examples
train_iterator = [
    ConversationExample(conversation=[
        "Hello",
        "Hi, how are you?",
        "Good, thank you, how about you?",
        "Good, thanks!"
    ]),
    ConversationExample(conversation=[
        "I am hungry",
        "How about eating pizza?"
    ]),
]
valid_iterator = [
    ConversationExample(conversation=[
        "Tired...",
        "Let's have a break!",
        "Nice idea!"
    ]),
]

# Fine-tune model
model.fit(train_iterator=train_iterator, valid_iterator=valid_iterator)

# Generate response
model.generate(context=["Hello", "How are you"], do_sample=True, top_p=0.95, top_k=50)
# Output could be like below if sufficient examples were given.
# => ConversationModelOutput(responses=['Good thank you'], context=['Hello', 'How are you'])
```

Please refer to [document](docs/en/docs/index.md) for more details of installation, model architecture and usage.

* [Install](docs/en/docs/install.md)
* [Model Architecture Overview](docs/en/docs/model_architecture_overview.md)
* [Model Training](docs/en/docs/model_training.md)
* [Response Generation](docs/en/docs/response_generation.md)
* [CLI (Experimental)](docs/en/docs/cli.md)

Enjoy talking with your conversational AI :wink:


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Add max_len option in `fit` method https://github.com/colorfulscoop/convmodel/pull/21
- Set default model path as a command line argument for `try` CLI https://github.com/colorfulscoop/convmodel/pull/24

### Fixed

- Add max_len option in Dataset to restrict the max length of a tensor inputting to a model https://github.com/colorfulscoop/convmodel/pull/20
- Revise code structure https://github.com/colorfulscoop/convmodel/pull/22
- Fix two issues in https://github.com/colorfulscoop/convmodel/pull/23
  - Responses include EOS at the end of the utterance.
  - Tokenizer adds special tokens if the tokenizer adds them as default.

## [v0.3.0]

### Added

- `output_path` and `save_best_model` parameters to `fit` method. https://github.com/colorfulscoop/convmodel/pull/13
  - `output_path`: when given, model is saved while training in the fit method. Default is None (will not be saved)
  - `save_best_model`: when given with output_path, fit will only save the best model in `output_path`
- `min_new_tokens` option to `generate` method. https://github.com/colorfulscoop/convmodel/pull/16
- eval method in model and cli. https://github.com/colorfulscoop/convmodel/pull/14
- `print_config` option to `fit` cli command. https://github.com/colorfulscoop/convmodel/pull/18

### Changed

- streamlit UI by using side-by-side columns to show conversation. https://github.com/colorfulscoop/convmodel/pull/15
- Change CLI command name from run_streamlit to try. https://github.com/colorfulscoop/convmodel/pull/17

## [v0.2.0]

### Added

- fit method in ConversatonModel. https://github.com/colorfulscoop/convmodel/pull/8
- fit CLI https://github.com/colorfulscoop/convmodel/pull/10

### Changed

- device transfer code in fit method. https://github.com/colorfulscoop/convmodel/pull/9
- document to use MkDocs. https://github.com/colorfulscoop/convmodel/pull/11

### Removed

- Obsoleted modules for bert and gpt2_lm. https://github.com/colorfulscoop/convmodel/pull/7/files

## [v0.1.1] - 2021-08-28

### Added

- build_data_loader method in ConversationDataset class. https://github.com/colorfulscoop/convmodel/pull/5

## [v0.1.0] - 2021-08-28

### Added

- ConversationModel and its usage.


convmodel/app.py
import streamlit as st
from convmodel.models.gpt2lm.model import GPT2LMConversationModel as ConversationModel
import os

st.set_page_config(
    page_title="Try your convmodel",
    # Set "wide" to use full screen
    layout="centered"
)


def show_conversation(placeholder, context):
    markdown = ""
    icon = ["🙂", "🤖"]
    for idx, item in enumerate(context):
        markdown += f"{icon[idx % 2]} {item}  \n"
    placeholder.markdown(markdown)


@st.cache
def load_model(model_dir):
    model = ConversationModel.from_pretrained(model_dir)
    return model


# Prepare context
if "context" not in st.session_state:
    st.session_state.context = []
# Prepare key for text input
if "text_input_key" not in st.session_state:
    st.session_state.text_input_key = 0


# Get default value from environment variable
MODEL_PATH = os.environ.get("MODEL_PATH", "model")

#
# Set sidebar
#
st.sidebar.markdown(
    """
    # Try your convmodel

    <small>convmodel documents available on [docs](https://colorfulscoop.github.io/convmodel-docs/).</small>

    <small>Check [Text Generation](https://huggingface.co/docs/transformers/main/en/main_classes/text_generation) document
    for the parameters of text generation.</small>

    ## Model Path
    """,
    unsafe_allow_html=True
)
model_dir = st.sidebar.text_input('Model', MODEL_PATH)

st.sidebar.markdown("## Text Generation Parameters", unsafe_allow_html=True)

do_sample = st.sidebar.checkbox('do_sample', value=False)
if do_sample:
    penalty_alpha = st.sidebar.number_input('penalty_alpha', value=0)
    top_p = st.sidebar.number_input('top_p', value=0.95)
    top_k = st.sidebar.number_input('top_k', value=50)
else:
    # Use contrastive search
    penalty_alpha = st.sidebar.number_input('penalty_alpha', value=0.6)
    top_p = st.sidebar.number_input('top_p', value=1.0)
    top_k = st.sidebar.number_input('top_k', value=4)

# max_length = st.sidebar.number_input('Max # of tokens to input to the model', value=1024)
min_new_tokens = st.sidebar.number_input('Min # of new tokens to generate', value=1)
max_new_tokens = st.sidebar.number_input('Max # of new tokens to generate', value=30)
num_beams = st.sidebar.number_input('Number of beams', value=1)
no_repeat_ngram_size = st.sidebar.number_input('no_repeat_ngram_size', value=0)
num_return_sequences = st.sidebar.number_input('Number of sentences generated', value=1)

#
# Set main layout
#

# Load model
try:
    model = load_model(model_dir)
    st.success(f'Loaded a model from the model path "{model_dir}" :wink:', icon="✅")
    st.markdown('Input your message below and press Enter to start conversation :speech_balloon:')
except OSError:
    st.warning(f'''
        Error happens while loading a model in "{model_dir}".

        Check the model is located in the directory. Error details are below.
    ''',
    icon="⚠️")
    raise
#st.info('Input your message and press Enter.')

# context area
precontext_str = st.text_area("Context")
precontext = [x.strip() for x in precontext_str.split("\n") if x.strip()]

col_left, col_right = st.columns(2)

# Left side
input_area = col_left.empty()
reset_button = col_left.button(
    'Reset conversation',
    # help="Reset conversation histroy to start new conversation."
)

if reset_button:
    st.session_state.context = []
    st.session_state.text_input_key += 1

user_input = input_area.text_input('User input', "", key=str(st.session_state.text_input_key))

# Right side
conversation_area = col_right.empty()
show_conversation(conversation_area, st.session_state.context)

if (not reset_button) and user_input:
    if not st.session_state.context:
        st.session_state.context.extend(precontext)

    # If user inputted a text, clea input text
    # This technique is explained below
    # https://github.com/streamlit/streamlit/issues/623#issuecomment-551755236
    st.session_state.text_input_key += 1
    input_area.text_input('User input', "", key=str(st.session_state.text_input_key))

    st.session_state.context.append(user_input)
    show_conversation(conversation_area, st.session_state.context)

    with st.spinner(text="(・ω・ ).oO(thinking...)"):
        gen = model.generate(
            context=st.session_state.context,
            do_sample=do_sample,
            top_p=top_p,
            top_k=top_k,
            # max_length=max_length,
            min_new_tokens=min_new_tokens,
            max_new_tokens=max_new_tokens,
            num_beams=num_beams,
            no_repeat_ngram_size=no_repeat_ngram_size,
            bad_words_ids=[[model.hf_tokenizer.unk_token_id]],
            num_return_sequences=num_return_sequences,
            penalty_alpha=penalty_alpha,
        )
        res = gen.responses[0]
        st.session_state.context.append(res)
        print(gen)

    show_conversation(conversation_area, st.session_state.context)
